File: js/src/selectors.js
import Op from './evm/opcodes.js'
import { Vm, UnsupportedOpError } from './evm/vm.js'
import { StackIndexError } from './evm/stack.js'
import Element from './evm/element.js'
import { toUint8Array, uint8ArrayToBigInt } from './utils.js'

function process(vm, gasLimit) {
  let selectors = new Set()
  let gasUsed = 0

  while (!vm.stopped) {
    // console.log('selectors', selectors)
    // console.log(vm.toString())
    let ret
    try {
      ret = vm.step()
      gasUsed += ret[1]
      if (gasUsed > gasLimit) {
        // throw `gas overflow: ${gasUsed} > ${gasLimit}`
        break
      }
    } catch (e) {
      if (e instanceof StackIndexError || e instanceof UnsupportedOpError) {
        // console.log(e)
        break
      } else {
        throw e
      }
    }

    const [op, , r0, r1] = ret

    switch (op) {
      case Op.XOR:
      case Op.EQ:
      case Op.SUB:
        if (r0.label === 'signature' || r1.label === 'signature') {
          selectors.add(uint8ArrayToBigInt(r0.label === 'signature' ? r1.data : r0.data))
          vm.stack.pop()
          vm.stack.push_uint(op === Op.EQ ? 0n : 1n)
        }
        break

      case Op.LT:
      case Op.GT:
        if (r0.label === 'signature' || r1.label === 'signature') {
          const clonedVm = vm.clone()
          const [newSelectors, gas] = process(clonedVm, Math.trunc((gasLimit - gasUsed) / 2))
          newSelectors.forEach((v) => selectors.add(v))
          gasUsed += gas
          const v = vm.stack.pop_uint()
          vm.stack.push_uint(v === 0n ? 1n : 0n)
        }
        break

      case Op.MUL:
        if (r0.label === 'signature' || r1.label === 'signature') {
          vm.stack.peek().label = 'mulsig'
        }
        break

      // Vyper _selector_section_dense()/_selector_section_sparse()
      // (sig MOD n_buckets) or (sig AND (n_buckets-1))
      case Op.MOD:
      case Op.AND:
        {
          const p = vm.stack.peek()
          if (
            (op === Op.AND && (r0.label === 'signature' || r1.label === 'signature')) ||
            (op === Op.MOD && (r0.label === 'mulsig' || r0.label === 'signature'))
          ) {
            const otd = op === Op.AND && r1.label == 'signature' ? r0.data : r1.data
            const rawMa = uint8ArrayToBigInt(otd)
            if (op === Op.AND && rawMa === 0xffffffffn) {
              p.label = 'signature'
            } else {
              if (rawMa < 256n) {
                const ma = Number(rawMa)
                vm.stack.pop()
                const to = op === Op.MOD ? ma : ma + 1
                for (let m = 1; m < to && gasUsed < gasLimit; m++) {
                  const clonedVm = vm.clone()
                  clonedVm.stack.push_uint(BigInt(m))
                  const [newSelectors, gas] = process(clonedVm, Math.trunc((gasLimit - gasUsed) / ma))
                  newSelectors.forEach((v) => selectors.add(v))
                  gasUsed += gas
                }
                vm.stack.push_uint(0n)
              }
            }
          } else if (op === Op.AND && (r0.label === 'calldata' || r1.label === 'calldata')) {
            p.label = 'calldata'
          }
        }
        break

      case Op.SHR:
        if (r1.label === 'calldata') {
          const p = vm.stack.peek()
          if (p.data.slice(-4).every((v, i) => v === vm.calldata.data[i])) {
            p.label = 'signature'
          } else {
            p.label = 'mulsig'
          }
        } else if (r1.label === 'mulsig') {
          const p = vm.stack.peek()
          p.label = 'mulsig'
        }
        break

      case Op.DIV:
        if (r0.label === 'calldata') {
          const p = vm.stack.peek()
          if (p.data.slice(-4).every((v, i) => v === vm.calldata.data[i])) {
            p.label = 'signature'
          }
        }
        break

      case Op.ISZERO:
        if (r0.label === 'signature') {
          selectors.add(0n)
        }
        break

      case Op.MLOAD:
        {
          const p = vm.stack.peek()
          if (r0.has('calldata') && p.data.slice(-4).every((v, i) => v === vm.calldata.data[i])) {
            p.label = 'signature'
          }
        }
        break
    }
  }
  return [selectors, gasUsed]
}

export function functionSelectors(code, gasLimit = 5e5) {
  const codeArr = toUint8Array(code)
  const vm = new Vm(codeArr, new Element(new Uint8Array([0xaa, 0xbb, 0xcc, 0xdd]), 'calldata'))
  const [selectors] = process(vm, gasLimit)
  return [...selectors.values()].map((x) => x.toString(16).padStart(8, '0'))
}

File: rust/src/selectors.rs
use crate::evm::{
    op,
    vm::{StepResult, Vm},
    Element, VAL_0_B, VAL_1_B,
};
use crate::Selector;

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
enum Label {
    CallData,
    Signature,
}

fn analyze(
    vm: &mut Vm<Label>,
    selectors: &mut Vec<Selector>,
    ret: StepResult<Label>,
    gas_used: &mut u32,
    gas_limit: u32,
) -> Result<(), Box<dyn std::error::Error>> {
    match ret {
          StepResult{op: op::XOR|op::EQ, fa: Some(Element{label: Some(Label::Signature), ..}), sa: Some(s1), ..}
        | StepResult{op: op::XOR|op::EQ, sa: Some(Element{label: Some(Label::Signature), ..}), fa: Some(s1), ..} =>
        {
            selectors.push(s1.data[28..32].try_into().unwrap());
            let v = vm.stack.peek_mut()?;
            v.data = if ret.op == op::XOR { VAL_1_B } else { VAL_0_B };
        }

          StepResult{op: op::SUB, fa: Some(Element{label: Some(Label::Signature), ..}), sa: Some(s1), ..}
        | StepResult{op: op::SUB, sa: Some(Element{label: Some(Label::Signature), ..}), fa: Some(s1), ..} =>
        {
            selectors.push(s1.data[28..32].try_into().unwrap());
        }

          StepResult{op: op::LT|op::GT, fa: Some(Element{label: Some(Label::Signature), ..}), ..}
        | StepResult{op: op::LT|op::GT, sa: Some(Element{label: Some(Label::Signature), ..}), ..} =>
        {
            *gas_used += process(vm.clone(), selectors, (gas_limit - *gas_used) / 2);
            let v = vm.stack.peek_mut()?;
            v.data = if v.data == VAL_0_B { VAL_1_B } else { VAL_0_B };
        }

          StepResult{op: op::SHR, sa: Some(Element{label: Some(Label::CallData), ..}), ..}
        | StepResult{op: op::AND, fa: Some(Element{label: Some(Label::Signature), ..}), ..}
        | StepResult{op: op::AND, sa: Some(Element{label: Some(Label::Signature), ..}), ..}
        | StepResult{op: op::DIV, fa: Some(Element{label: Some(Label::CallData), ..}), ..} =>
        {
            let v = vm.stack.peek_mut()?;
            if v.data[28..32] == vm.calldata.data[0..4] {
                v.label = Some(Label::Signature);
            }
        }

          StepResult{op: op::AND, fa: Some(Element{label: Some(Label::CallData), ..}), ..}
        | StepResult{op: op::AND, sa: Some(Element{label: Some(Label::CallData), ..}), ..} =>
        {
            let v = vm.stack.peek_mut()?;
            v.label = Some(Label::CallData);
        }

        StepResult{op: op::ISZERO, fa: Some(Element{label: Some(Label::Signature), ..}), ..} =>
        {
            selectors.push([0; 4]);
        }

        StepResult{op: op::MLOAD, ul: Some(used), ..} =>
        {
            if used.contains(&Label::CallData) {
                let v = vm.stack.peek_mut()?;
                v.label = Some(
                    if v.data[28..32] == vm.calldata.data[0..4] {
                        Label::Signature
                    } else {
                        Label::CallData
                    }
                )
            }
        }

        _ => {}
    }
    Ok(())
}

fn process(mut vm: Vm<Label>, selectors: &mut Vec<Selector>, gas_limit: u32) -> u32 {
    let mut gas_used = 0;
    while !vm.stopped {
        // println!("{:?}\n", vm);
        let ret = match vm.step() {
            Ok(v) => v,
            Err(_e) => {
                // eprintln!("{}", _e);
                break;
            }
        };
        gas_used += ret.gas_used;
        if gas_used > gas_limit {
            break;
        }

        if analyze(&mut vm, selectors, ret, &mut gas_used, gas_limit).is_err() {
            break;
        }
    }
    gas_used
}

/// Extracts function selectors
///
/// # Arguments
///
/// * `code` - A slice of deployed contract bytecode
/// * `gas_limit` - Maximum allowed gas usage; set to `0` to use defaults
///
/// # Examples
///
/// ```
/// use evmole::function_selectors;
/// use hex::decode;
///
/// let code = decode("6080604052348015600e575f80fd5b50600436106030575f3560e01c80632125b65b146034578063b69ef8a8146044575b5f80fd5b6044603f3660046046565b505050565b005b5f805f606084860312156057575f80fd5b833563ffffffff811681146069575f80fd5b925060208401356001600160a01b03811681146083575f80fd5b915060408401356001600160e01b0381168114609d575f80fd5b80915050925092509256").unwrap();
///
/// let selectors: Vec<_> = function_selectors(&code, 0);
///
/// assert_eq!(selectors, vec![[0x21, 0x25, 0xb6, 0x5b], [0xb6, 0x9e, 0xf8, 0xa8]])
/// ```
pub fn function_selectors(code: &[u8], gas_limit: u32) -> Vec<Selector> {
    let vm = Vm::<Label>::new(
        code,
        Element {
            data: [0xaa, 0xbb, 0xcc, 0xdd, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            label: Some(Label::CallData),
        },
    );
    let mut selectors: Vec<Selector> = Vec::new();
    process(
        vm,
        &mut selectors,
        if gas_limit == 0 {
            5e5 as u32
        } else {
            gas_limit
        },
    );
    selectors
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_code() {
        let s = function_selectors(&[], 0);
        assert_eq!(s.len(), 0);
    }
}
